File: bench/spectral-norm/go/spectral-norm.go
// spectral-norm — see ../c/spectral-norm.c for the algorithm.
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

func a(i, j int) float64 {
	s := i + j
	return 1.0 / float64(s*(s+1)/2+i+1)
}
func mulAv(v, out []float64) {
	n := len(v)
	for i := 0; i < n; i++ {
		s := 0.0
		for j := 0; j < n; j++ {
			s += a(i, j) * v[j]
		}
		out[i] = s
	}
}
func mulAtv(v, out []float64) {
	n := len(v)
	for i := 0; i < n; i++ {
		s := 0.0
		for j := 0; j < n; j++ {
			s += a(j, i) * v[j]
		}
		out[i] = s
	}
}
func mulAtAv(v, out, tmp []float64) {
	mulAv(v, tmp)
	mulAtv(tmp, out)
}
func main() {
	n := 100
	if len(os.Args) > 1 {
		if v, err := strconv.Atoi(os.Args[1]); err == nil {
			n = v
		}
	}
	u := make([]float64, n)
	v := make([]float64, n)
	tmp := make([]float64, n)
	for i := range u {
		u[i] = 1.0
	}
	for i := 0; i < 10; i++ {
		mulAtAv(u, v, tmp)
		mulAtAv(v, u, tmp)
	}
	vBv, vv := 0.0, 0.0
	for i := 0; i < n; i++ {
		vBv += u[i] * v[i]
		vv += v[i] * v[i]
	}
	fmt.Printf("%.9f\n", math.Sqrt(vBv/vv))
}

File: bench/spectral-norm/python/spectral-norm.py
# spectral-norm — see ../c/spectral-norm.c for the algorithm.
import sys
import math


def a(i, j):
    s = i + j
    return 1.0 / (s * (s + 1) // 2 + i + 1)


def mul_Av(v, out):
    n = len(v)
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += a(i, j) * v[j]
        out[i] = s


def mul_Atv(v, out):
    n = len(v)
    for i in range(n):
        s = 0.0
        for j in range(n):
            s += a(j, i) * v[j]
        out[i] = s


def mul_AtAv(v, out, tmp):
    mul_Av(v, tmp)
    mul_Atv(tmp, out)


def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    u = [1.0] * n
    v = [0.0] * n
    tmp = [0.0] * n
    for _ in range(10):
        mul_AtAv(u, v, tmp)
        mul_AtAv(v, u, tmp)
    vBv = vv = 0.0
    for i in range(n):
        vBv += u[i] * v[i]
        vv += v[i] * v[i]
    print("%.9f" % math.sqrt(vBv / vv))


main()
